src/Features/Core/Portable/UseCoalesceExpression/AbstractUseCoalesceExpressionDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,111 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.LanguageServices;
+
+namespace Microsoft.CodeAnalysis.UseCoalesceExpression
+{
+    internal abstract class AbstractUseCoalesceExpressionDiagnosticAnalyzer<
+        TSyntaxKind,
+        TExpressionSyntax,
+        TConditionalExpressionSyntax,
+        TBinaryExpressionSyntax> : AbstractCodeStyleDiagnosticAnalyzer
+        where TSyntaxKind : struct
+        where TExpressionSyntax : SyntaxNode
+        where TConditionalExpressionSyntax : TExpressionSyntax
+        where TBinaryExpressionSyntax : TExpressionSyntax
+    {
+        protected AbstractUseCoalesceExpressionDiagnosticAnalyzer() 
+            : base(IDEDiagnosticIds.UseCoalesceExpressionDiagnosticId,
+                   new LocalizableResourceString(nameof(FeaturesResources.Use_coalesce_expression), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+        }
+
+        protected abstract TSyntaxKind GetSyntaxKindToAnalyze();
+        protected abstract ISyntaxFactsService GetSyntaxFactsService();
+        protected abstract bool IsEquals(TBinaryExpressionSyntax condition);
+        protected abstract bool IsNotEquals(TBinaryExpressionSyntax condition);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(AnalyzeSyntax, GetSyntaxKindToAnalyze());
+        }
+
+        private void AnalyzeSyntax(SyntaxNodeAnalysisContext context)
+        {
+            var conditionalExpression = (TConditionalExpressionSyntax)context.Node;
```

Comment:
Not sure many people care anymore, but this could be restricted to C# 2 and above.

It would be interesting to add to AbstractCodeStyleDiagnosticAnalyzer a way to restrict the language version.